current/runtime/storage.md

```diff
@@ -4,49 +4,237 @@ lang: en
+  Mutate the value with the provided function.
+- [`take()`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageValue.html#tymethod.take) -
+  Load the value and remove it from storage.
+
+### Storage Maps
+
+Map data structures are ideal for managing sets of items whose elements will be accessed randomly,
+as opposed to iterating over them sequentially in their entirety. Storage Maps in Substrate are
+implemented as key-value hash maps, which is a pattern that should be familiar to most developers.
+In order to give blockchain engineers increased control, Substrate allows developers to select
+[the hashing algorithm](#Hashing-Algorithms) that is used to generate a map's keys. Refer to
+[the advanced storage documentation](../advanced/storage) to learn more about how Substrate's
+Storage Maps are implemented.
+
+#### Methods
+
+[Storage Maps expose an API](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#required-methods)
+that is similar to that of Storage Values.
+
+- `get` - Load the value associated with the provided key from storage. Docs:
+  [`StorageMap#get(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.get),
+  [`StorageDoubleMap#get(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.get)
+- `insert` - Store the provided value by associating it with the given key. Docs:
+  [`StorageMap#insert(key, val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.insert),
+  [`StorageDoubleMap#insert(key1, key2, val)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.insert)
+- `mutate` - Use the provided function to mutate the value associated with the given key. Docs:
+  [`StorageMap#mutate(key, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.mutate),
+  [`StorageDoubleMap#mutate(key1, key2, fn)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.mutate)
+- `take` - Load the value associated with the given key and remove it from storage. Docs:
+  [`StorageMap#take(key)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageMap.html#tymethod.take),
+  [`StorageDoubleMap#take(key1, key2)`](https://substrate.dev/rustdocs/master/frame_support/storage/trait.StorageDoubleMap.html#tymethod.take)
+
+#### Iterable Storage Maps
+
+Depending on [the hashing algorithm](#Transparent-Hashing-Algorithms) that you select to generate a
+map's keys, you may be able to iterate across its keys and values. Because maps are often used to
+track unbounded sets of data (account balances, for example) it is especially likely to exceed block
+production time by iterating over maps in their entirety within the runtime. Furthermore, because
```

Comment:
> Furthermore, because
accessing the elements of a map requires more pointer dereferencing than accessing the elements of a
native list, maps are significantly _more_ costly than lists to iterate over with respect to time.

Can you explain this a bit more? I am not sure where it comes from and until now my understanding of the state was that the most important factors encoding + db access. In-memory dereferencing seems negligible to me.

Reply:
(I agree that iterating maps is expensive, I am not sure if your reason is correct.)

Reply:
So, this line used to read as follows:
> Furthermore, because maps are comprised of more layers of indirection than native lists...

I agree with you that the change didn't necessarily make things more clear. I guess what I meant above was that you don't need to dereference _memory_ pointers, per se, but you do have to dereference the pointer to the element in the map...I think I'm trying to describe the "db access" part that you mention above. I will think about alternate ways to phrase this but would definitely appreciate any suggestions :+1:

Reply:
Okay, this now reads:
> ...accessing the elements of a map requires more database reads than accessing the elements of a native list...

I hope you agree the simplest approach was the best one in this case :sweat_smile: